### PR TITLE
Adding Ramsete control

### PR DESCRIPTION
**Pull request by mxgordon**
This still needs some polishing, but I'm making it now so people can review it

### CODE REVIEW

**Comment by mxgordon** [comment]
I believe this also add some arm stuff cause I merged from there. btw this is the branch that was used at bunnybots

File: auto/AutoPath.java [hunk 1]
```diff
@@ -0,0 +1,32 @@
...
+     * @param fileName name of file (everything before the extension)
+     */
+    public AutoPath(String fileName) {
+        filePath = "paths/output/" + fileName + ".wpilib.json";
+        path = Filesystem.getDeployDirectory().toPath().resolve(filePath);
+
+        try {
+            this.trajectory = TrajectoryUtil.fromPathweaverJson(path);
+        } catch (IOException e) {
+            // I want this to fail hard to people know quickly when something is wrong instead of a NullPointerException down the road
```
Comment (iscruzrivera) [inline auto/AutoPath.java]: * so. also, please remove any personal pronouns from your documentation

File: auto/RamseteCommandWrapper.java [hunk 2]
```diff
@@ -0,0 +1,220 @@
...
+        NetworkTableEntry rightReference = SmartDashboard.getEntry("right_reference");
+        NetworkTableEntry rightMeasurement = SmartDashboard.getEntry("right_measurement");
+
+        RamseteController ramseteController = new RamseteController();
+
+        PIDController left = new PIDController(AutoConstants.KP, 0, 0);
+        PIDController right = new PIDController(AutoConstants.KP, 0, 0);
+
+        if (debuggingMode) {
+            // This "disables" the fancy control from the ramsete controller, allow you to
```
Comment (iscruzrivera) [inline auto/RamseteCommandWrapper.java]: why isn't this a multi-line comment

File: auto/RamseteCommandWrapper.java [hunk 3]
```diff
@@ -0,0 +1,220 @@
...
+            // verify the feedforward and PID gains. It is the same as doing what is
+            // suggested here:
+            // https://docs.wpilib.org/en/stable/docs/software/advanced-controls/trajectories/troubleshooting.html#verify-feedforward
+            // * you can also comment this line out if you don't want to disable the controller
+            ramseteController.setEnabled(false);
+
+            // * To test feedforward gains, uncomment these following lines (and make sure
+            // debugging mode is on)
```
Comment (iscruzrivera) [inline auto/RamseteCommandWrapper.java]: add an enter after this line so it's clear what you're uncommenting and what you're not

**Review by csokolove — APPROVED** [review]

File: controllers/ArmFPIDController.java [hunk 4]
```diff
@@ -0,0 +1,97 @@
...
+    public ArmFPIDController(double f, double p, double i, double d, ArmFeedforward armFF) {
+        super(p, i, d);
+        this.f = f;
+        this.p = p;
+        this.i = i;
+        this.d = d;
+        this.armFF = armFF;
+    }
+
+    // returns true if Feedforward is enabled
```
Comment (iscruzrivera) [inline controllers/ArmFPIDController.java]: the documentation here should follow java docs convention. you can find a link to that here: https://www.tutorialspoint.com/java/java_documentation.htm
this applies to all of the comments in this file

File: controllers/FFController.java [hunk 5]
```diff
@@ -0,0 +1,79 @@
...
+    public double calculateElevator(double velocity, double acceleration) {
+        return kS * Math.signum(velocity) + kG + kV * velocity + kA * acceleration;
+    }
+
+    public double calculateElevator(double velocity) {
+        return calculateElevator(velocity, 0);
+    }
+
+    /**  
+    * use this for an elevator, or any FPIDSubsystem that does not require an angle gain
```
Comment (iscruzrivera) [inline controllers/FFController.java]: you have a comment here that says to use this for an elevator, but then proceed to have arm methods?

File: subsystems/FPIDSubsystem.java [hunk 6]
```diff
@@ -0,0 +1,185 @@
...
+    public static class FPIDConstants {
+        public double speedMultiplier, gearing;
+        public double[] softLimits;
+        public int encoderCPR, zeroOffset, switchPort;
+        public boolean switchInverted;
+        public PID pid;
+        public FFController ff;
+        public TrapezoidProfile.Constraints constraints;
+
+        public FPIDConstants(
```
Comment (iscruzrivera) [inline subsystems/FPIDSubsystem.java:24]: uhhh why is this formatted onto so many lines

Comment (iscruzrivera) [reply subsystems/FPIDSubsystem.java:24]: use the vscode short hand command to clean this file up

File: subsystems/FPIDSubsystem.java [hunk 7]
```diff
@@ -0,0 +1,185 @@
...
+        super(
+            new ProfiledPIDController(
+                constants.pid.kP, 
+                constants.pid.kI, 
+                constants.pid.kD,
+                constants.constraints
+            )
+        );
+        
+        // this.enable();
```
Comment (iscruzrivera) [inline subsystems/FPIDSubsystem.java]: why is this commented out - delete it if it's unecessary

File: subsystems/FPIDSubsystem.java [hunk 8]
```diff
@@ -0,0 +1,185 @@
...
+    }
+
+    /**
+    * Have the implementing class configure the motors 
+    */
+    public abstract void configureMotors();
+
+    /**
+    * the implementing class also needs to get and set encoder values, as we don't know the encoder type
+    */
```
Comment (iscruzrivera) [inline subsystems/FPIDSubsystem.java]: add like an empty line between this and the next part to make it clearer

File: subsystems/SimpleMotorBase.java [hunk 9]
```diff
@@ -19,11 +20,11 @@ public SimpleMotorBase(SpeedController motor, double forwardThrottle, double bac
...
-        motorController.set(forwardThrottle);
+        motorGroup.set(forwardThrottle);
     }
 
     public void backward() {
-        motorController.set(-backwardThrottle);
+        motorGroup.set(-backwardThrottle);
     }
 
     public void stop() {
```
Comment (iscruzrivera) [inline subsystems/SimpleMotorBase.java:30]: for the line under it, shouldn't it be motorGroup.set(0)?

**Review by iscruzrivera — CHANGES_REQUESTED** [review]

**Comment by csokolove** [comment]
I started working on the Javadocs for ArmFPIDController file, but I held it from the last commit only cause there are a couple methods I'm unsure of the exact functionality, so I'm gonna let whoever did those methods finish those docs.

I'll push the documentation I did that I am sure of though

EDIT:
Pushed it.
* Someone else needs to do the param & return descriptions for the methods after line 70 that don't have them
* Someone also needs to do documentation for the `calculate` method (line 103 at time of writing)

File: auto/AutoPath.java [hunk 10]
```diff
@@ -0,0 +1,32 @@
...
+     * @param fileName name of file (everything before the extension)
+     */
+    public AutoPath(String fileName) {
+        filePath = "paths/output/" + fileName + ".wpilib.json";
+        path = Filesystem.getDeployDirectory().toPath().resolve(filePath);
+
+        try {
+            this.trajectory = TrajectoryUtil.fromPathweaverJson(path);
+        } catch (IOException e) {
+            // Fail hard to people know quickly when something is wrong instead of a NullPointerException down the road
```
Comment (iscruzrivera) [inline auto/AutoPath.java]: * so

File: controllers/ArmFPIDController.java [hunk 11]
```diff
@@ -0,0 +1,129 @@
...
+import edu.wpi.first.wpilibj.controller.PIDController;
+import edu.wpi.first.wpilibj.smartdashboard.SendableBuilder;
+import edu.wpi.first.wpilibj.controller.ArmFeedforward;
+
+public class ArmFPIDController extends PIDController {
+    protected double f, p, i, d;
+    protected ArmFeedforward armFF;
+    private boolean fDisabled = false;
+
+    // params fpid are Feedforward and PID gains, param sCosVA is an array of the ks, kcos, kv, and ka used by an ArmFeedForward
```
Comment (iscruzrivera) [inline controllers/ArmFPIDController.java]: this stuff should be javadocs too

**Review by iscruzrivera — APPROVED** [review]

**Review by iscruzrivera — APPROVED** [review]

**Pull request merged**